### PR TITLE
Add `Encoder`/`BuildEncoder` and `TransformerEncoder` implementation

### PR DESCRIPTION
**Pull request by danieldk**

### CODE REVIEW

File: src/architectures/encoder.rs [hunk 1]
```diff
@@ -13,26 +14,77 @@ pub struct EncoderOutput {
...
+    /// * `type_ids` - Input type ids.
+    ///   *Shape:* `(batch_size, seq_len)`
+    /// * `train` - Whether to train the layer.
+    fn forward_t(
+        &self,
+        piece_ids: &Tensor,
+        attention_mask: &AttentionMask,
+        positions: Option<&Tensor>,
+        type_ids: Option<&Tensor>,
+        train: bool,
```
Comment (shadeMe) [inline src/architectures/encoder.rs:58]: `train` [should come before](https://github.com/oxidized-transformers/oxidized-transformers/blob/main/CODING.md#module-forward-signature) the optional arguments, I think.

File: src/architectures/encoder.rs [hunk 2]
```diff
@@ -46,8 +98,8 @@ pub trait EncoderLayer {
     /// *Shape:* ``(batch_size, seq_len, width)``
     fn forward_t(
         &self,
-        piece_ids: &Tensor,
-        attention_mask: &AttentionMask,
+        input: &Tensor,
+        mask: &AttentionMask,
         positions: Option<&Tensor>,
         train: bool,
```
Comment (shadeMe) [inline src/architectures/encoder.rs:104]: Here as well (but maybe better scoped in another PR?).

File: src/models/transformer/encoder.rs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        for layer in &self.layers {
+            let next_layer_output = layer
+                .forward_t(&layer_output, attention_mask, positions, train)
+                .context(TransformerLayerSnafu)?;
+
+            layer_outputs.push(next_layer_output.clone());
+            layer_output = next_layer_output;
+        }
+
+        if let Some(last) = layer_outputs.last_mut() {
```
Comment (shadeMe) [inline src/models/transformer/encoder.rs:152]: Can we avoid the `if let Some(...)` seeing how `layer_outputs` will always have at least one element?

**Review by shadeMe — CHANGES_REQUESTED** [review]

File: src/architectures/encoder.rs (hunk 1 above)
Comment (danieldk) [reply src/architectures/encoder.rs:58]: Yeah, as you suggest, it's probably better for me to do this in a separate PR, so I can go through my dev branch first and all the existing models on the branch get updated through the refactor.

File: src/models/transformer/encoder.rs (hunk 3 above)
Comment (danieldk) [reply src/models/transformer/encoder.rs:152]: Not sure if it's really cleaner and barely shorter (same when we use indexing)?

```
        let last = layer_outputs.last_mut().unwrap();
        *last = self
            .output_layer_norm
            .forward_t(last, train)
            .context(LayerNormSnafu)?;
```

Comment (shadeMe) [reply src/models/transformer/encoder.rs:152]: Yeah, not much of a difference.

**Review by shadeMe — APPROVED** [review]

**Pull request merged**